### PR TITLE
simplify block subscriber and add check/log block checks

### PR DESCRIPTION
**Pull request by FelixFan1992**

### CODE REVIEW

File: core/web/presenters/job.go [hunk 1]
```diff
@@ -196,6 +196,7 @@ type OffChainReporting2Spec struct {
 	CreatedAt                         time.Time              `json:"createdAt"`
 	UpdatedAt                         time.Time              `json:"updatedAt"`
 	CollectTelemetry                  bool                   `json:"collectTelemetry"`
+	CollectAutomationCustomTelemetry  bool                   `json:"collectAutomationCustomTelemetry"`
```
Comment (infiloop2) [inline core/web/presenters/job.go:199]: nit. unrelated change

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go [hunk 2]
```diff
@@ -18,107 +18,96 @@ import (
...
+	// channelSize represents the channel size for head broadcaster
+	channelSize = 20
 )
 
-type BlockKey struct {
+type blockKey struct {
 	block int64
-	hash  common.Hash
+	hash  string
 }
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:30]: not sure if we need this struct. Can we directly use `ocr2keepers.BlockKey` in its place?

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go [hunk 3]
```diff
@@ -18,107 +18,96 @@ import (
...
-	// PollLogInterval is the interval to query log poller
-	PollLogInterval = time.Second
-	// CleanUpInterval is the interval for cleaning up block maps
-	CleanUpInterval = 15 * time.Minute
-	// ChannelSize represents the channel size for head broadcaster
-	ChannelSize = 20
+	// cleanUpInterval is the interval for cleaning up block maps
+	cleanUpInterval = 15 * time.Minute
+	// channelSize represents the channel size for head broadcaster
+	channelSize = 20
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:24]: what does this indicate? Is it the max blocks stored in the channel via head broadcaster while it waits to be read by this component?

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go [hunk 4]
```diff
@@ -18,107 +18,96 @@ import (
...
+		blocks:           map[int64]string{},
+		blockHistorySize: blockHistorySize,
+		lggr:             lggr.Named("BlockSubscriber"),
 	}
 }
 
-func (hw *BlockSubscriber) getBlockRange(ctx context.Context) ([]uint64, error) {
-	h, err := hw.lp.LatestBlock(pg.WithParentCtx(ctx))
+func (hw *BlockSubscriber) getBlockRange() ([]uint64, error) {
+	h, err := hw.lp.LatestBlock(pg.WithParentCtx(hw.ctx))
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:70]: do we need the log poller dependency here? thought we were trying to remove it

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go [hunk 5]
```diff
@@ -18,107 +18,96 @@ import (
...
+		subscribers:      map[int]chan ocr2keepers.BlockHistory{},
+		blocks:           map[int64]string{},
+		blockHistorySize: blockHistorySize,
+		lggr:             lggr.Named("BlockSubscriber"),
 	}
 }
 
-func (hw *BlockSubscriber) getBlockRange(ctx context.Context) ([]uint64, error) {
-	h, err := hw.lp.LatestBlock(pg.WithParentCtx(ctx))
+func (hw *BlockSubscriber) getBlockRange() ([]uint64, error) {
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:69]: not sure i follow the motivation for removing ctx argument from here

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go [hunk 6]
```diff
@@ -18,107 +18,96 @@ import (
...
 	}
 	return blocks, nil
 }
 
-func (hw *BlockSubscriber) getLogPollerBlocks(ctx context.Context, blocks []uint64) error {
-	// request the past LOOK_BACK blocksFromPoller from log poller
-	// returned blocksFromPoller are in ASC order
-	logpollerBlocks, err := hw.lp.GetBlocksRange(ctx, blocks, pg.WithParentCtx(ctx))
+func (hw *BlockSubscriber) initializeBlocks(blocks []uint64) error {
+	logpollerBlocks, err := hw.lp.GetBlocksRange(hw.ctx, blocks, pg.WithParentCtx(hw.ctx))
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:84]: still using log poller?

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go (hunk 3 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:24]: yes

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go (hunk 4 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:70]: the log poller is only used when this service starts. the reason is i am concerned that when this service just starts, if subscriber needs block history for 128 or 256 blocks, the head broadcaster's depth can only trace back to finality depth.
if this is not a serious issue, we can drop log poller dependency completely.

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go (hunk 5 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:69]: since hw struct already has a ctx so i am just using it directly

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go (hunk 2 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:30]: removed

File: core/web/presenters/job.go (hunk 1 above)
Comment (FelixFan1992) [reply core/web/presenters/job.go:199]: removed

**Comment by FelixFan1992** [comment]
This PR aims to simplify how block subscriber works.
Currently, it obtains blocks from 2 sources: log poller (lp) and head broadcaster (hb). It will make db queries to lp every 1 second to pull 128 blocks and listen to block head from hb. This adds on complexity and workload to the subscriber and lp. The reason to do this is to protect against re-org.
However, we can have a better and simpler way to protect against re-org. `types.Head` from hb has a pointer to its parent `types.Head`. When re-org happens, the new block will have a pointer to new parent from re-org, which also has new parent. Tracing back this pointer, we can get all the new blocks (at most EVM finality depth) from re-org.
Because we need 128 blocks to build history and when this service starts, new heads from hb only have EVM finality depth blocks. It will take a while to build up 128 blocks. Hence, the log poller is still kept as a dependency to obtain 128 block history when the service starts.

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go [hunk 7]
```diff
@@ -141,26 +117,40 @@ func (hw *BlockSubscriber) Start(_ context.Context) error {
 	return hw.sync.StartOnce("BlockSubscriber", func() error {
 		hw.mu.Lock()
 		defer hw.mu.Unlock()
-		_, hw.unsubscribe = hw.hb.Subscribe(&headWrapper{headC: hw.headC})
 		hw.ctx, hw.cancel = context.WithCancel(context.Background())
```
Comment (amirylm) [inline core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:120]: use the context argument (currently ignored) instead of `context.Background`

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go (hunk 5 above)
Comment (amirylm) [reply core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:57]: Using a context like that is a bit hacky, let's use the one we get as a parameter (although we will pass the same context probably)

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go [hunk 8]
```diff
@@ -141,26 +117,40 @@ func (hw *BlockSubscriber) Start(_ context.Context) error {
...
+						hw.lggr.Infof("blocks block %d hash is %s", h.Number, h.Hash.Hex())
 
-						history := hw.buildHistory(bk.block)
+						history := hw.buildHistory(h.Number)
 
-						hw.lastSentBlock = bk.block
+						hw.lastSentBlock = h.Number
 						hw.lggr.Infof("lastSentBlock is %d", hw.lastSentBlock)
 						// send history to all subscribers
 						for _, subC := range hw.subscribers {
```
Comment (amirylm) [inline core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:156]: Writing on channels might block, wrap with a select to ensure that we won't get affected by some subscriber:
```
select {
  case subC <- history:
  default:
    hw.lggr.Warnf("subscriber channel is full, dropping block history with length %d", len(history))
}
```

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go (hunk 7 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:120]: done

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go [hunk 9]
```diff
@@ -141,26 +117,40 @@ func (hw *BlockSubscriber) Start(_ context.Context) error {
...
 			go func(ctx context.Context) {
 				for {
 					select {
-					case bk := <-hw.headC:
+					case h := <-hw.headC:
 						hw.mu.Lock()
-						if hw.lastClearedBlock == 0 {
-							hw.lastClearedBlock = bk.block - 1
-							hw.lggr.Infof("lastClearedBlock is %d", hw.lastClearedBlock)
+						// head parent is a linked list with EVM finality depth
```
[the remark this answers is not in the document]
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:141]: done

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go (hunk 8 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:156]: done

Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:156]: good idea

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go (hunk 5 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go]: done

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go [hunk 10]
```diff
@@ -261,23 +210,50 @@ func (hw *BlockSubscriber) Unsubscribe(subId int) error {
...
+	defer hw.mu.Unlock()
+	// head parent is a linked list with EVM finality depth
+	// when re-org happens, new heads will have pointers to the new blocks
+	for cp := h; cp != nil; cp = cp.Parent {
+		if cp != h && hw.blocks[cp.Number] != cp.Hash.Hex() {
+			hw.lggr.Warnf("overriding block %d old hash %s with new hash %s due to re-org", cp.Number, hw.blocks[cp.Number], cp.Hash.Hex())
+		}
+		hw.blocks[cp.Number] = cp.Hash.Hex()
+	}
+	hw.lggr.Infof("blocks block %d hash is %s", h.Number, h.Hash.Hex())
```
Comment (amirylm) [inline core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:224]: change to `Debug`

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go [hunk 11]
```diff
@@ -261,23 +210,50 @@ func (hw *BlockSubscriber) Unsubscribe(subId int) error {
...
+	for _, subC := range hw.subscribers {
+		// wrapped in a select to not get blocked by certain subscribers
+		select {
+		case subC <- history:
+		default:
+			hw.lggr.Warnf("subscriber channel is full, dropping block history with length %d", len(history))
+		}
+	}
+
+	hw.lggr.Infof("published block history with length %d to %d subscriber(s)", len(history), len(hw.subscribers))
```
Comment (amirylm) [inline core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:241]: change to `Debug` and add latest block to log?

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go [hunk 12]
```diff
@@ -926,3 +937,76 @@ func (r *EvmRegistry) getBlockHash(blockNumber *big.Int) (common.Hash, error) {
...
+		UpkeepID:               p.UpkeepID,
+		Trigger:                p.Trigger,
+		WorkID:                 p.WorkID,
+		FastGasWei:             big.NewInt(0),
+		LinkNative:             big.NewInt(0),
+	}
+}
+
+// verifyCheckBlock checks that the check block and hash are valid
+func (r *EvmRegistry) verifyCheckBlock(checkBlock, upkeepId *big.Int, checkHash common.Hash, p ocr2keepers.UpkeepPayload, i int, results []ocr2keepers.CheckResult) bool {
```
Comment (amirylm) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:963]: return error and set the result from the caller

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go [hunk 13]
```diff
@@ -18,55 +19,43 @@ import (
...
+	hb               httypes.HeadBroadcaster
+	lp               logpoller.LogPoller
+	headC            chan *evmtypes.Head
+	unsubscribe      func()
+	subscribers      map[int]chan ocr2keepers.BlockHistory
+	blocks           map[int64]string
+	maxSubId         int
+	lastClearedBlock int64
+	lastSentBlock    int64
+	latestBlock      *atomic.Int64
```
[the remark this answers is not in the document]
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:42]: done

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go (hunk 10 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:224]: done

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go (hunk 12 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:963]: done

**Comment by FelixFan1992** [comment]
this PR does a few things:
1. simplify block subscriber to mostly use head broadcaster
2. verify check block / hash
3. verify log block / hash (verify tx hash if log block and hash not available)

File: core/services/ocr2/plugins/ocr2keeper/evm21/abi.go [hunk 14]
```diff
@@ -13,23 +13,30 @@ import (
...
+	UpkeepFailureReasonLogBlockInvalid
+	UpkeepFailureReasonTxHashNoLongerExists
+
+	// pipeline execution error
+	NoPipelineError     = 0
+	CheckBlockTooOld    = 1
+	CheckBlockInvalid   = 2
+	RpcFlakyFailure     = 3
+	MercuryFlakyFailure = 4
+	PackUnpackFailed    = 5
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/abi.go:39]: nit. maybe a new enum would be better here?

File: core/services/ocr2/plugins/ocr2keeper/evm21/abi.go [hunk 15]
```diff
@@ -13,23 +13,30 @@ import (
...
-	UPKEEP_FAILURE_REASON_UPKEEP_NOT_NEEDED
-	UPKEEP_FAILURE_REASON_PERFORM_DATA_EXCEEDS_LIMIT
-	UPKEEP_FAILURE_REASON_INSUFFICIENT_BALANCE
-	UPKEEP_FAILURE_REASON_MERCURY_CALLBACK_REVERTED
-	UPKEEP_FAILURE_REASON_REVERT_DATA_EXCEEDS_LIMIT
-	UPKEEP_FAILURE_REASON_REGISTRY_PAUSED
-
-	// Start of offchain failure types. All onchain failure reasons from
-	// contract should be put above
-	UPKEEP_FAILURE_REASON_MERCURY_ACCESS_NOT_ALLOWED
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/abi.go:32]: can we still keep this split? I think it is still valid?

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go [hunk 16]
```diff
@@ -261,23 +210,49 @@ func (hw *BlockSubscriber) Unsubscribe(subId int) error {
 	return nil
 }
 
+func (hw *BlockSubscriber) processHead(h *evmtypes.Head) {
+	hw.mu.Lock()
+	defer hw.mu.Unlock()
+	// head parent is a linked list with EVM finality depth
+	// when re-org happens, new heads will have pointers to the new blocks
+	for cp := h; cp != nil; cp = cp.Parent {
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:218]: nit. are we guaranteed to reach a nil parent after finality depth? Maybe safer to also exit after initialBlockSize iterations?

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go [hunk 17]
```diff
@@ -38,18 +38,21 @@ import (
...
-	// CleanupInterval decides when the expired items in cache will be deleted.
-	CleanupInterval = 25 * time.Minute
+	// defaultAllowListExpiration decides how long an upkeep's allow list info will be valid for.
+	defaultAllowListExpiration = 20 * time.Minute
+	// cleanupInterval decides when the expired items in cache will be deleted.
+	cleanupInterval = 25 * time.Minute
+	// LookbackDepth decides valid trigger block lookback range
+	LookbackDepth = 1024
+	// BlockHistorySize decides the block history size
+	BlockHistorySize = 128
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:48]: I think these make more sense to be defined as constants within block_subscriber.go instead of passing as arguments from here

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go [hunk 18]
```diff
@@ -926,3 +936,82 @@ func (r *EvmRegistry) getBlockHash(blockNumber *big.Int) (common.Hash, error) {
 
 	return block.Hash(), nil
 }
+
+func (r *EvmRegistry) getTxBlock(txHash common.Hash) (*big.Int, common.Hash, error) {
+	txr, err := r.client.TransactionReceipt(r.ctx, txHash)
+	if err != nil {
+		return nil, common.Hash{}, err
+	}
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:944]: do we know what client.TransactionReceipt returns when the tx is not found? Does it return an empty txr object or an error?

We need to differentiate here b/w flaky errors and tx not found

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go [hunk 19]
```diff
@@ -926,3 +936,82 @@ func (r *EvmRegistry) getBlockHash(blockNumber *big.Int) (common.Hash, error) {
...
+
+	return txr.BlockNumber, txr.BlockHash, nil
+}
+
+func (r *EvmRegistry) getCheckResult(p ocr2keepers.UpkeepPayload, reason, state uint8, retryable bool) ocr2keepers.CheckResult {
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:949]: nit. This seems like a checkResult without performData, i.e. a result generated due to some error or the upkeep being ineligible. Can we clarify that in the name somehow?

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go [hunk 20]
```diff
@@ -926,3 +936,82 @@ func (r *EvmRegistry) getBlockHash(blockNumber *big.Int) (common.Hash, error) {
...
+	// if log block number is populated, check log block number and block hash
+	if logBlockNumber != 0 {
+		h, ok := r.bs.queryBlocksMap(logBlockNumber)
+		if ok && h == logBlockHash.Hex() {
+			r.lggr.Debugf("tx hash %s exists on chain at block number %d for upkeepId %s", hexutil.Encode(p.Trigger.LogTriggerExtension.TxHash[:]), logBlockNumber, upkeepId)
+			return UpkeepFailureReasonNone, false
+		}
+		r.lggr.Warnf("log block %d does not exist in block subscriber for upkeepId %s, querying eth client", logBlockNumber, upkeepId)
+	} else {
+		r.lggr.Warnf("log block not provided, querying eth client for tx hash %s for upkeepId %s", hexutil.Encode(p.Trigger.LogTriggerExtension.TxHash[:]), upkeepId)
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:1003]: nit. i think we should log this without the else bracket too, i.e. also when we were not able to get logBlockNumber within bs

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go [hunk 21]
```diff
@@ -926,3 +936,82 @@ func (r *EvmRegistry) getBlockHash(blockNumber *big.Int) (common.Hash, error) {
...
+	}
+	// query eth client as a fallback
+	bn, _, err := r.getTxBlock(p.Trigger.LogTriggerExtension.TxHash)
+	if err != nil {
+		r.lggr.Warnf("failed to query tx hash %s for upkeepId %s: %s", hexutil.Encode(p.Trigger.LogTriggerExtension.TxHash[:]), upkeepId, err.Error())
+		return UpkeepFailureReasonLogBlockInvalid, true
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:1009]: does this mean a retryable error?

File: core/services/ocr2/plugins/ocr2keeper/evm21/abi.go (hunk 15 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/abi.go:32]: done

File: core/services/ocr2/plugins/ocr2keeper/evm21/abi.go (hunk 14 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/abi.go:39]: done

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go (hunk 16 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:218]: done

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go (hunk 17 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:48]: done

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go (hunk 19 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:949]: done

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go (hunk 21 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:1009]: yes

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go (hunk 20 above)
Comment (FelixFan1992) [reply core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:1003]: you mean log this outside else?

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go [hunk 22]
```diff
@@ -18,266 +19,249 @@ import (
...
-			hash:  head.BlockHash(),
-		}
-	}
-}
-
-func getBlockHistory(keys []BlockKey) ocr2keepers.BlockHistory {
-	var blockKeys []ocr2keepers.BlockKey
-	for _, k := range keys {
-		blockKeys = append(blockKeys, k.getBlockKey())
+		w.headC <- head
```
Comment (amirylm) [inline core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:265]: wrap with select to avoid blocking

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go [hunk 23]
```diff
@@ -18,266 +19,252 @@ import (
...
+	h, err := bs.lp.LatestBlock(pg.WithParentCtx(ctx))
 	if err != nil {
 		return nil, err
 	}
-	hw.lggr.Infof("latest block from log poller is %d", h)
+	bs.lggr.Infof("latest block from log poller is %d", h)
 
 	var blocks []uint64
-	for i := int64(0); i < hw.blockHistorySize; i++ {
+	for i := bs.blockSize - 1; i >= 0; i-- {
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:74]: check for h-i > 0

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go [hunk 24]
```diff
@@ -18,266 +19,252 @@ import (
...
-			keys = append(keys, BlockKey{
-				block: block - i,
-				hash:  h1,
-			})
-		} else if h2, ok2 := hw.blocksFromBroadcaster[block-i]; ok2 {
-			// if a block only exists in broadcaster, use data from broadcaster
-			keys = append(keys, BlockKey{
-				block: block - i,
-				hash:  h2,
+	for i := int64(0); i < bs.blockHistorySize; i++ {
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:99]: check for block-i > 0

File: core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go [hunk 25]
```diff
@@ -18,266 +19,252 @@ import (
...
-						return
-					}
-				}
-			}(hw.ctx)
+func (bs *BlockSubscriber) Start(ctx context.Context) error {
+	bs.lggr.Info("block subscriber started.")
+	return bs.sync.StartOnce("BlockSubscriber", func() error {
+		bs.mu.Lock()
+		defer bs.mu.Unlock()
+		bs.ctx, bs.cancel = context.WithCancel(ctx)
```
Comment (infiloop2) [inline core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go:129]: use new context here

File: core/services/ocr2/plugins/ocr2keeper/evm21/registry.go (hunk 18 above)
Comment (infiloop2) [reply core/services/ocr2/plugins/ocr2keeper/evm21/registry.go:940]: let's add a todo till then to merge

**Review by amirylm — APPROVED** [review]
